### DISCUSSION TITLE
GButton protected members

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#1983] [General]
Hi,

Can we turn two protected members as public?
https://github.com/axmolengine/axmol/blob/6304c88e14736a8b78777a3fd10f6acfdef1ed38/extensions/fairygui/src/fairygui/GButton.h#L67-L68

I can't change control state manually using gamepad controller without access it.

This line is making the `setSelected` unusable, since it start in COMMON state when initialize.

```cpp
void GButton::setSelected(bool value)
{
    if (_mode == ButtonMode::COMMON)
        return;
```

Ref: https://github.com/axmolengine/axmol/blob/6304c88e14736a8b78777a3fd10f6acfdef1ed38/extensions/fairygui/src/fairygui/GButton.cpp#L103-L106

The common mode is the default button behavior, but i can't simulate hover/selected effect without have access to update it state:

<img width="206" alt="image" src="https://github.com/axmolengine/axmol/assets/395096/9ff34f8b-a336-449c-8926-f025f836328b">

Thanks.

**Comment by paulocoutinhox** [comment]
My temporary workaround:

```cpp
#include "gui/MenuWindowGUI.h"

namespace game::gui {

USING_NS_FGUI;

class GButton2 : public GButton {
public:
    void xyz() {
        setState(GButton::SELECTED_OVER);
    }
};

void MenuWindowGUI::onInit() {
    setContentPane(UIPackage::createObject("Main", "GameMenuWindow")->as<GComponent>());
    center();

    getContentPane()->getChildByPath("btStart")->addClickListener(CC_CALLBACK_1(MenuWindowGUI::onBtStartClick, this));
}

void MenuWindowGUI::onShown() {
    auto button =static_cast<GButton2 *>(getContentPane()->getChildByPath("btStart"));
    button->xyz();
}

void MenuWindowGUI::onBtStartClick(fairygui::EventContext *context) {
    btStartClickCallback(context);
}

} // namespace game::gui
```

**Reply by smilediver** [reply]
I don't think modifying 3rd party lib or its API is a good idea.

What type of button are you implementing? A simple clickable button or a checkbox like thing? As I understand, the hover is done using `GButton::onRollOver()` and `GButton::onRollOut()`. `GButton::setSelected()` is used for checkboxes, not buttons.

**Reply by paulocoutinhox** [reply]
What you use to change the state manually?

**Reply by smilediver** [reply]
We don't use FairyGUI, I just read the code and docs.

**Reply by paulocoutinhox** [reply]
What you use for games UI creation?

**Reply by smilediver** [reply]
We use our custom UI systems.
